src/profile-logic/profile-data.js

```diff
@@ -1522,3 +1522,111 @@ export function getTreeOrderComparator(
     return compareCallNodes(callNodeA, callNodeB);
   };
 }
+
+/**
+ * This function walks up the from a clicked call node, and tries to find the best
+ * call node to select. This is the root-most call node that is drawn on the same
+ * category path as the clicked call node.
+ */
+export function findBestAncestorCallNode(
+  callNodeInfo: CallNodeInfo,
+  sampleCallNodes: Array<IndexIntoCallNodeTable | null>,
+  sampleCategories: Array<IndexIntoCategoryList | null>,
+  clickedCallNode: IndexIntoCallNodeTable,
+  clickedCategory: IndexIntoCategoryList
+): IndexIntoCallNodeTable {
+  const { callNodeTable } = callNodeInfo;
+  if (callNodeTable.category[clickedCallNode] !== clickedCategory) {
+    return clickedCallNode;
+  }
+
+  // Compute the callNodesOnSameCategoryPath.
+  // Given a call node path with some arbitrary categories, e.g. A, B, C
+  //
+  // A -> A -> B -> B -> C -> C -> C
+  //
+  // This loop will select the leaf-most call nodes that match the leaf call-node's
+  // category. Running the above path through this loop would produce the list:
+  //
+  // "C -> C -> C".
+  const callNodesOnSameCategoryPath = [clickedCallNode];
+  let callNode = clickedCallNode;
+  while (true) {
+    const parentCallNode = callNodeTable.prefix[callNode];
+    if (parentCallNode === -1) {
+      // The entire call path is just clickedCategory.
+      return clickedCallNode; // TODO: is this a useful behavior?
+    }
+    if (callNodeTable.category[parentCallNode] !== clickedCategory) {
+      break;
+    }
+    callNodesOnSameCategoryPath.push(parentCallNode);
+    callNode = parentCallNode;
+  }
+
+  // Now find the callNode in callNodesOnSameCategoryPath with the lowest depth
+  // such that selecting it will not highlight any samples whose unfiltered
+  // category is different from clickedCategory. If no such callNode exists,
+  // return clickedCallNode.
+
+  const clickedDepth = callNodeTable.depth[clickedCallNode];
+  const handledCallNodes = new Uint8Array(callNodeTable.length);
+  function limitSameCategoryPathToCommonAncestor(callNode) {
```

Comment:
Add a comment here: `callNode` is the leaf call node of a sample whose sample category is a different category than `clickedCategory`. If `callNode`'s ancestor path crosses `callNodesOnSameCategoryPath`, that implies that `callNode` would be highlighted if we were to select the root-most node in `callNodesOnSameCategoryPath`. If that is the case, we need to truncate `callNodesOnSameCategoryPath` in such a way that the root-most node in that list is no longer an ancestor of `callNode`.